swo_rql/query_builder.py

```diff
@@ -0,0 +1,364 @@
+from swo_rql import constants
+
+
+def parse_kwargs(query_dict):
+    query = []
+    for lookup, value in query_dict.items():
+        tokens = lookup.split("__")
+        if len(tokens) == 1:
+            #  field=value
+            field = tokens[0]
+            value = rql_encode("eq", value)
+            query.append(f"eq({field},{value})")
+            continue
+        op = tokens[-1]
+        if op not in constants.KEYWORDS:
+            # field__nested=value
+            field = ".".join(tokens)
+            value = rql_encode("eq", value)
+            query.append(f"eq({field},{value})")
+            continue
+        field = ".".join(tokens[:-1])
+        if op in constants.COMP or op in constants.SEARCH:
+            value = rql_encode(op, value)
+            query.append(f"{op}({field},{value})")
+            continue
+        if op in constants.LIST:
+            value = rql_encode(op, value)
+            query.append(f"{op}({field},({value}))")
+            continue
+
+        cmpop = "eq" if value is True else "ne"
+        expr = "null()" if op == constants.NULL else "empty()"
+        query.append(f"{cmpop}({field},{expr})")
+
+    return query
+
+
+def rql_encode(op, value):
+    from datetime import date, datetime
+    from decimal import Decimal
+
+    if op not in constants.LIST:
+        if isinstance(value, str):
+            return value
+        if isinstance(value, bool):
+            return "true" if value else "false"
+        if isinstance(value, int | float | Decimal):
+            return str(value)
+        if isinstance(value, date | datetime):
+            return value.isoformat()
+    if op in constants.LIST and isinstance(value, list | tuple):
+        return ",".join(value)
+    raise TypeError(f"the `{op}` operator doesn't support the {type(value)} type.")
+
+
+class RQLQuery:
+    """
+    Helper class to construct complex RQL queries.
+
+    Usage:
+
+    ```py3
+        rql = R(field='value', field2__in=('v1', 'v2'), field3__empty=True)
+    ```
+    !!! note
+        All the lookups expressed as keyword arguments are combined together with a logical `and`.
+
+
+    Using the ``n`` method:
+
+    ```py3
+        rql = (
+            R().n('field').eq('value')
+            & R().n('field2').anyof(('v1', 'v2'))
+            & R().n('field3').empty(True)
+        )
+    ```
+
+    The previous query can be expressed in a more concise form like:
+
+    ```py3
+    rql = R().field.eq('value') & R().field2.anyof(('v1', 'v2')) & r.field3.empty(True)
```

Comment:
That syntax looks really nice for me.

Espacially if one time we would like to have models, so you can put in `R` model and later operate on this model with the guarantee to check if field is available or not

```
rql = R(Model).field.eq()
```